app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/casiogb6900/CasioGATTThread.java

```diff
@@ -0,0 +1,48 @@
+/*      Copyright (C) 2018 Andreas Böhler
+        based on code from BlueWatcher, https://github.com/masterjc/bluewatcher
+
+        This file is part of Gadgetbridge.
+
+        Gadgetbridge is free software: you can redistribute it and/or modify
+        it under the terms of the GNU Affero General Public License as published
+        by the Free Software Foundation, either version 3 of the License, or
+        (at your option) any later version.
+
+        Gadgetbridge is distributed in the hope that it will be useful,
+        but WITHOUT ANY WARRANTY; without even the implied warranty of
+        MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+        GNU Affero General Public License for more details.
+
+        You should have received a copy of the GNU Affero General Public License
+        along with this program.  If not, see <http://www.gnu.org/licenses/>. */
+package nodomain.freeyourgadget.gadgetbridge.service.devices.casiogb6900;
+import android.content.Context;
+
+public class CasioGATTThread extends Thread {
+    CasioGATTServer mServer = null;
+
+    public CasioGATTThread(Context context, CasioGB6900DeviceSupport deviceSupport)
+    {
+        mServer = new CasioGATTServer(context, deviceSupport);
+    }
+
+    public void setContext(Context ctx)
+    {
+        mServer.setContext(ctx);
+    }
+
+    @Override
+    public void run()
+    {
+        mServer.initialize();
+        while(true)
+        {
+            try {
+                wait(100);
+            } catch(Exception e)
+            {
+
+            }
+        }
```

Comment:
Right. I tried running the server without this thread, but then the client communication stalls. I suppose that the thread is terminated when run() finishes, that's why I added the calls to sleep(). I should've added a way to stop the thread, though.

Reply:
Yes, the thread should terminate together with the *DeviceSupport instance. I guess from there you should call the new quit() method? BTW, for calling `wait()` you need to `synchronize` on `this` (see the method's javadoc).

And where did you get the time to wait from (100ms)? Waking up every 100ms sounds a bit battery intensive to me.

For the gatt client (which most other devices use), we use the `BtLEQueue` to handle all gatt interactions in a separate thread with proper synchronization and ordering. Maybe we need the same for gatt *servers*!?

Do you have some time and energy to investigate how the gatt server communication is supposed to work?
- with proper synchronization
- without waiting/polling

I don't want to scare you away with such a big task -- I'm really happy about your contribution already!

The thing is that getting the lowlevel communication right is essential to getting a reliable connection and maintaining low resource usage, considering different phones, different OS versions, different BLE drivers, and especially different gadgets. Even more so because such code tends to get "reused" elsewhere sooner or later.

Reply:
I'm going to use an object to wait and increase the wait time to a minute or so. The 100ms were based on the time it then takes until the thread quits if it's just checking the flag. With the object, I can notify the thread prematurely. I'll push the updated code as soon as it's ready.

Regarding the queue: IMHO that's a separate task after the PR has been accepted. I'm not sure how much time I can invest in the coming weeks, but I'm quite positive that I can find some spare time for this. 

Should we create an issue to discuss the implementation options? I'm sure I'll have some more questions on the inner workings of Gadgetbridge.

Reply:
Yeah, I'm fine with doing this separately. I'll open a ticket for this.